Add checks for overlapping TupleTypes

elazarg opened the pull request:
Fix #4237.

This is not complete or bulletproof, but it's an improvement.

I don't see where should there be a special case for namedtuple; maybe there shouldn't.

JukkaL commented on mypy/meet.py:
```diff
@@ -120,6 +120,17 @@ class C(A, B): ...
+            if all(is_overlapping_types(ti, si, use_promotions)
+                   for ti, si in zip(t.items, s.items)):
+                return True
+        # TODO: Tuple[A, ...]
```
I think that this may work already -- `Tuple[A, ...]` is represented as an `Instance`, and the new check below should handle this (`TupleType` vs. `Instance`). Can you add a test case and remove this comment if I was correct?

elazarg replied:
The case was not handled well, so I added support for it

elazarg commented:
Added support for `Tuple[A, ...]`.
I also fixed the use of fallback - I believe that comparing fallbacks is incorrect. 

@JukkaL the logic is quite different now, so it needs to be re-reviewed... Sorry about that :\

There are _many_ tests that can be added - involving combinations of tuples, namedtuple, varargs, instances and unions; but I prefer to do it in at a different time.
